### PR TITLE
feat(http): add support for JSONP requests

### LINKED ISSUE #2818: feat(http): add JsonP and JsonPBackend
Create a new subclass of `Http` called `JsonP`, which uses a new `JsonPBackend` to perform requests using [JsonP](https://en.wikipedia.org/wiki/JSONP) techniques. 

The callback name to provide in the URL should be `JSON_CALLBACK`, similar to AngularJS 1.x.

The `JsonP` implementation should throw if asked to perform the request with a method other "GET". 

This should have its own `injectables` so that users who don't need it don't get the extra bytes.

### PR DESCRIPTION
**Pull request by caitp**
Dart is broken, WIP

Closes #2818

### CODE REVIEW

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            .forEach(method => {
+              let base = new BaseRequestOptions();
+              let req = new Request(
+                  base.merge(new RequestOptions({url: 'https://google.com', method: method})));
+              expect(() => new JSONPConnection(req, new MockBrowserJsonp())).toThrow();
+            });
+      });
+
+      it('should respond with data passed to callback', inject([AsyncTestCompleter], async => {
+           let connection = new JSONPConnection(sampleRequest, new MockBrowserJsonp());
```
Comment (caitp) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: I'm not going to manually fix clang-format errors

File: modules/angular2/src/facade/lang.dart [hunk 2]
```diff
@@ -246,4 +252,4 @@ class DateWrapper {
 }
 
 // needed to match the exports from lang.js
-var global = null;
+var global = context;
```
Comment (caitp) [inline modules/angular2/src/facade/lang.dart]: Importing `dart:js` seems to cause problems with `gulp build.dart` --- but this isn't going to work without that. I don't think it works with `dart:html`'s `window` either, but I'll try that.

Comment (caitp) [inline modules/angular2/src/facade/lang.dart]: nope, nothing related to dart ever works, no good

**Comment by caitp** [comment]
anyone who knows anything about dart have an idea here? the JS code is happy

**Comment by jeffbcross** [comment]
@caitp The latest travis build [#6453](https://travis-ci.org/angular/angular/builds/69792100) looks like it's just choking on TypeScript errors. Can you point to a build where dart errors were apparent?

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts (hunk 1 above)
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: When you run gulp enforce-format, it should give you the script to run to automatically fix all clang complaints.

File: modules/angular2/src/facade/lang.dart (hunk 2 above)
Comment (jeffbcross) [inline modules/angular2/src/facade/lang.dart]: What are the problems when importing dart:js, and what do you need from dart:js?

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+           ObservableWrapper.subscribe(connection.response, res => {
+             expect(res.json()).toEqual(({fake_payload: true, blob_id: 12345})) async.done();
```
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: Fixing this missing semicolon, and moving async.done to the next line will get rid of all ts2dart errors.

**Comment by jeffbcross** [comment]
I've started addressing some of the dart issues on my fork: https://github.com/jeffbcross/angular/tree/pr-2905

I'll get back into it tomorrow and will connect with you on the right approach to get this working in dart, which is still not completely clear to me. Requires a little more research into dart.

**Comment by jeffbcross** [comment]
Yeah, I think my fork is not the direction to go. We'll most likely just need to create a `browser_jsonp.dart` similar to how I did with `browser_xhr`, rather than use the DomAdapter facade like I was trying to do.

**Comment by caitp** [comment]
PTAL --- it's working locally for me on dart2js and production js, and passing all cjs tests

**Comment by caitp** [comment]
actually, a slight fixup was needed for cjs that broke the dart build --- it seems dart2js doesn't handle "protected" properties

**Comment by caitp** [comment]
Just blocked on https://github.com/angular/ts2dart/pull/240 I guess

**Comment by jeffbcross** [comment]
I merged angular/ts2dart#240, but will have to revisit Monday because there's a release process for ts2dart that I'm not familiar with.

**Comment by jeffbcross** [comment]
Though you could update the PR to use latest ts2dart package if you want to see if Travis goes green.

**Comment by caitp** [comment]
I'll add a commit to test that tonight just in case the patch wasn't enough

**Comment by caitp** [comment]
well that didn't work

**Comment by caitp** [comment]
okay, so it works now, but the build fails because of dart benchmarks trying to use `JitProtoChangeDetector` --- not sure why that changed

**Comment by caitp** [comment]
I'm not sure why the benchmark uses `!isJsObject({})` rather than just `!JitChangeDetection.isSupported()`?

**Comment by jeffbcross** [comment]
Whoa, Travis is actually green! Is this ready for review, @caitp?

**Comment by caitp** [comment]
yes, by all means =)

File: modules/angular2/src/facade/lang.dart [hunk 4]
```diff
@@ -190,6 +190,10 @@ class BaseException extends Error {
   }
 }
 
+Error makeTypeError([String message = ""]) {
+  return new BaseException(message);
```
Comment (jeffbcross) [inline modules/angular2/src/facade/lang.dart]: Why not TypeError? https://api.dartlang.org/apidocs/channels/be/dartdoc-viewer/dart:core.TypeError

File: modules/angular2/src/http/backends/browser_jsonp.dart [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+class BrowserJsonp {
+  // Construct a <script> element with the specified URL
+  dynamic build(String url) {
+    var node = document.createElement('script');
+    node.src = url;
+    return node;
+  }
+
+  nextRequestID() {
+    return "__req${_nextRequestId++}";
```
Comment (jeffbcross) [inline modules/angular2/src/http/backends/browser_jsonp.dart]: I was just looking at the AngularJS httpBackend implementation and found that the id is toStringed as base 36: https://github.com/angular/angular.js/commit/fd38655e6c8e58e7c7325b2a5937b26174ebdefa#diff-725ec6acb6c38b9cec7c30861f35be8cR31

Do you have any idea what that's for?

File: modules/angular2/src/http/http.ts [hunk 6]
```diff
@@ -113,10 +113,9 @@ export class Http {
   request(url: string | Request, options?: IRequestOptions): EventEmitter {
     var responseObservable: EventEmitter;
     if (isString(url)) {
-      responseObservable = httpRequest(
-          this._backend,
-          new Request(mergeOptions(this._defaultOptions, options, RequestMethods.GET, url)));
-    } else if (url instanceof Request) {
+      url = new Request(mergeOptions(this._defaultOptions, options, RequestMethods.GET, url));
```
Comment (jeffbcross) [inline modules/angular2/src/http/http.ts]: Should use _defaultOptions.method instead of RequestMethods.GET

File: modules/angular2/src/facade/lang.dart (hunk 4 above)
Comment (caitp) [inline modules/angular2/src/facade/lang.dart]: Dart's TypeError class doesn't let me attach an arbitrary message. I guess Dart doesn't use it as a catch-all exception the way JS does

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      ]);
+      backend = injector.get(JSONPBackend);
+      let base = new BaseRequestOptions();
+      sampleRequest = new Request(base.merge(new RequestOptions({url: 'https://google.com'})));
+    });
+
+    afterEach(() => { existingScripts = []; });
+
+    it('should create a connection',
+       () => { expect(() => backend.createConnection(sampleRequest)).not.toThrow(); });
```
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: Could you add an instanceof check?

File: modules/angular2/src/facade/lang.dart (hunk 4 above)
Comment (jeffbcross) [inline modules/angular2/src/facade/lang.dart]: Gotcha.

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+           existingScripts[0].dispatchEvent('load');
+         }));
+
+      it('should ignore load/callback when disposed', inject([AsyncTestCompleter], async => {
+           var connection = new JSONPConnection(sampleRequest, new MockBrowserJsonp());
+           connection.dispose();
+           expect(connection.readyState).toBe(ReadyStates.CANCELLED);
+
+           ObservableWrapper.subscribe(connection.response, res => {
+             expect("UNREACHABLE CODE").not.toBe("REACHED");
```
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: I don't understand this expectation. It would pass under any circumstance because you're comparing two string literals.

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+             async.done();
+           });
+           connection.finished('Fake data');
+           existingScripts[0].dispatchEvent('load');
+
```
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: Couldn't this cause expectations to be skipped?

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+           TimerWrapper.setTimeout(() => async.done(), 10);
```
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: Couldn't this cause expectations to be skipped?

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+         }));
+
+      it('should report error if loaded without invoking callback',
+         inject([AsyncTestCompleter], async => {
+           let connection = new JSONPConnection(sampleRequest, new MockBrowserJsonp());
+
+           ObservableWrapper.subscribe(
+               connection.response,
+               res => {
+                 expect("UNREACHABLE CODE").not.toBe("REACHED");
```
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: Same as above.

File: npm-shrinkwrap.json [hunk 12]
```diff
@@ -15385,36 +15385,36 @@
       }
     },
     "ts2dart": {
-      "version": "0.6.8",
-      "from": "https://registry.npmjs.org/ts2dart/-/ts2dart-0.6.8.tgz",
-      "resolved": "https://registry.npmjs.org/ts2dart/-/ts2dart-0.6.8.tgz",
+      "version": "0.6.9",
+      "from": "git://github.com/caitp/ts2dart.git#0.6.9",
```
Comment (jeffbcross) [inline npm-shrinkwrap.json]: I or someone needs to release ts2dart, and then you'd need to update shrinkwrap again.

**Comment by jeffbcross** [comment]
Left some minor comments, mostly around tests being written in a way where expectations could be skipped, or aren't valid,

I'll cut the ts2dart release this afternoon, and with the comments addressed, should be good to merge. I'll wait to give it green label until ts2dart is released.

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts (hunk 10 above)
Comment (caitp) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: in this case, I think a flake is extremely unrealistic --- but what else can we do?

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts (hunk 8 above)
Comment (caitp) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: yeah, the wording makes sense but in practice it doesn't really work. We do need some kind of UNREACHABLE() assertion though

File: modules/angular2/src/http/http.ts (hunk 6 above)
Comment (caitp) [inline modules/angular2/src/http/http.ts]: I plan to just revert the change to the Http portion entirely, it doesn't belong here. There's another bug related to it

File: modules/angular2/src/http/backends/browser_jsonp.dart (hunk 5 above)
Comment (caitp) [inline modules/angular2/src/http/backends/browser_jsonp.dart]: extra complexity? Igor might have a convincing reason for it

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts (hunk 7 above)
Comment (caitp) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: will do

File: npm-shrinkwrap.json (hunk 12 above)
Comment (caitp) [inline npm-shrinkwrap.json]: That's fine --- can we get that done today? It looks like you just need to run the build and `npm publish ./` locally

**Comment by caitp** [comment]
yeah, will make the suggested changes

**Comment by caitp** [comment]
Those commits should address the review comments, still waiting on the ts2dart release so that I can pull that in here. I'll squash when checking in

**Comment by jeffbcross** [comment]
TS2Dart has been updated: https://github.com/angular/angular/commit/3810e4bed3747eaf5d9f6cdd4d989657cff16e71

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts (hunk 10 above)
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: I think I'm misunderstanding this line. Isn't it essentially saying "complete the test in 10 seconds"? But shouldn't the test either complete inside the observable onNext function, or time out?

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts (hunk 1 above)
Comment (caitp) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: The script is what produced the bad sources, out just looks really really bad

File: modules/angular2/test/http/backends/jsonp_backend_spec.ts (hunk 10 above)
Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: Oh, now I see what's going on. The expectations are not meant to be evaluated, and you're waiting 10ms to give them a chance to fail.

Comment (jeffbcross) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: It would be clearer if this test used a spy, and inside the timeout verify that the spy hadn't been called. Outside of this PR, we should improve our stream implementation to allow immediate subscription to prevent the need for this async trickery.

Comment (caitp) [inline modules/angular2/test/http/backends/jsonp_backend_spec.ts]: No observable method is fired when a jsonp connection is disposed, so the assertion is that the response onNext is never invoked. It has a few ms to test this, which is potentially flaky. Another option is to enjoy some other notification on cancellation

**Comment by jeffbcross** [comment]
LGTM! Merge when Travis goes green.
